### PR TITLE
Add a serialized variant of the Trie-router

### PR DESCRIPTION
**Pull request by Joannis**
Since this trie is contiguous memory, routing performance should improve. This does not implement recursive wildcards yet, and I found that those tests are missing too.

In a quick test setup, the gap between no-router and having the old trie router was ~3-4k req/sec on a baseline (with router) of 98k. That gap shrunk by roughly 1k req/sec, leading to an estimated improvement of 25% in a small routing table (PerformanceBenchmarks).

I expect, and seemed to find a more significant performance gap when as the amount of path components increased. The expectation is based on the fact that more trie nodes are created when there are more path components. By keeping the routing table contiguous, it's expected that these scenarios are more performant.

APIs tend to have more path components to organise themselves, with many backends opting for `/api/v1`-type of prefixes for versioning. `/api/v1/users/:id/profile` is not a very long route.

## TODO

- [x] Proper Benchmarks
- [x] Add a Recursive Wildcard test
- [x] Add a recursive wildcard implementation

### CODE REVIEW

**Comment by Joannis** [comment]
## Benchmark Results
```
BinaryTrieRouter
╒═══════════════════════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╕
│ Metric                    │      p0 │     p25 │     p50 │     p75 │     p90 │     p99 │    p100 │ Samples │
╞═══════════════════════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╡
│ Malloc (total) *          │      24 │      24 │      24 │      24 │      24 │      24 │      24 │     164 │
├───────────────────────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Throughput (# / s) (K)    │     167 │     166 │     165 │     164 │     162 │     158 │     157 │     164 │
├───────────────────────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Time (total CPU) (ns) *   │    6001 │    6038 │    6058 │    6091 │    6169 │    6332 │    6353 │     164 │
╘═══════════════════════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╛

BinaryTrieRouterParameters
╒═══════════════════════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╕
│ Metric                    │      p0 │     p25 │     p50 │     p75 │     p90 │     p99 │    p100 │ Samples │
╞═══════════════════════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╡
│ Malloc (total) *          │      28 │      28 │      28 │      28 │      28 │      28 │      28 │     145 │
├───────────────────────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Throughput (# / s) (K)    │     148 │     148 │     147 │     145 │     142 │     138 │     138 │     145 │
├───────────────────────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Time (total CPU) (ns) *   │    6768 │    6795 │    6820 │    6906 │    7033 │    7201 │    7210 │     145 │
╘═══════════════════════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╛

Middleware Chain
╒═══════════════════════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╕
│ Metric                    │      p0 │     p25 │     p50 │     p75 │     p90 │     p99 │    p100 │ Samples │
╞═══════════════════════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╡
│ Malloc (total) *          │    1052 │    2657 │    3165 │    3777 │    4355 │    5483 │    6271 │    1112 │
├───────────────────────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Throughput (# / s) (#)    │    1599 │    1535 │    1508 │    1467 │    1385 │    1212 │    1079 │    1112 │
├───────────────────────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Time (total CPU) (μs) *   │     797 │     831 │     846 │     871 │     931 │    1102 │    1280 │    1112 │
╘═══════════════════════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╛

Router:Echo
╒═══════════════════════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╕
│ Metric                    │      p0 │     p25 │     p50 │     p75 │     p90 │     p99 │    p100 │ Samples │
╞═══════════════════════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╡
│ Malloc (total) *          │     853 │    2067 │    2397 │    2897 │    3565 │    4739 │    5452 │     877 │
├───────────────────────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Throughput (# / s) (#)    │    1162 │    1129 │    1115 │    1088 │    1024 │     898 │     767 │     877 │
├───────────────────────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Time (total CPU) (μs) *   │    1068 │    1122 │    1143 │    1170 │    1262 │    1568 │    1691 │     877 │
╘═══════════════════════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╛

Router:GET
╒═══════════════════════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╕
│ Metric                    │      p0 │     p25 │     p50 │     p75 │     p90 │     p99 │    p100 │ Samples │
╞═══════════════════════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╡
│ Malloc (total) *          │     612 │    2177 │    2563 │    3193 │    4009 │    5223 │    6929 │    1212 │
├───────────────────────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Throughput (# / s) (#)    │    1798 │    1714 │    1682 │    1636 │    1563 │    1391 │    1227 │    1212 │
├───────────────────────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Time (total CPU) (μs) *   │     722 │     762 │     778 │     799 │     847 │     994 │    1173 │    1212 │
╘═══════════════════════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╛

Router:PUT
╒═══════════════════════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╕
│ Metric                    │      p0 │     p25 │     p50 │     p75 │     p90 │     p99 │    p100 │ Samples │
╞═══════════════════════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╡
│ Malloc (total) *          │    1985 │    2949 │    3293 │    3689 │    4131 │    5079 │    6128 │     722 │
├───────────────────────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Throughput (# / s) (#)    │     928 │     894 │     878 │     851 │     802 │     681 │     592 │     722 │
├───────────────────────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Time (total CPU) (μs) *   │    1281 │    1337 │    1363 │    1413 │    1525 │    1969 │    2280 │     722 │
╘═══════════════════════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╛

TrieRouter
╒═══════════════════════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╕
│ Metric                    │      p0 │     p25 │     p50 │     p75 │     p90 │     p99 │    p100 │ Samples │
╞═══════════════════════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╡
│ Malloc (total) *          │      64 │      64 │      64 │      64 │      64 │      64 │      64 │     110 │
├───────────────────────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Throughput (# / s) (K)    │     112 │     111 │     110 │     109 │     107 │     105 │     105 │     110 │
├───────────────────────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Time (total CPU) (ns) *   │    8966 │    9003 │    9069 │    9191 │    9347 │    9454 │    9514 │     110 │
╘═══════════════════════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╛

TrieRouterParameters
╒═══════════════════════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╕
│ Metric                    │      p0 │     p25 │     p50 │     p75 │     p90 │     p99 │    p100 │ Samples │
╞═══════════════════════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╡
│ Malloc (total) *          │      78 │      78 │      78 │      78 │      78 │      78 │      78 │      90 │
├───────────────────────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Throughput (# / s) (K)    │      91 │      90 │      90 │      90 │      89 │      84 │      84 │      90 │
├───────────────────────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Time (total CPU) (μs) *   │      11 │      11 │      11 │      11 │      11 │      12 │      12 │      90 │
╘═══════════════════════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╛
```

**Comment by adam-fowler** [comment]
Can you do a benchmark comparison, with the original

**Comment by Joannis** [comment]
@adam-fowler the comparison is between `TrieRouter` / `TrieRouterParameters` and `BinaryTrieRouter` / `BinaryTrieRouterParameter`

**Comment by Joannis** [comment]
I've reduces the bloat from my benchmark comment

**Comment by adam-fowler** [comment]
This is failing CI

File: Sources/Hummingbird/Router/BinaryTrie/BinaryTrie+resolve.swift [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+        case .deadEnd:
+            return .deadEnd
+        }
+    }
+
+    /// A function that takes a path component and descends the trie to find the value
+    private func descendPath(
+        in trie: inout ByteBuffer,
+        index: UInt16,
+        parameters: inout Parameters,
```
Comment (adam-fowler) [inline Sources/Hummingbird/Router/BinaryTrie/BinaryTrie+resolve.swift:143]: Given you return the parameters, why is this marked `inout`?

File: Sources/Hummingbird/Router/BinaryTrie/BinaryTrie+resolve.swift [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+#elseif canImport(Musl)
+import Musl
+#elseif os(Linux) || os(FreeBSD) || os(Android)
+import Glibc
+#else
+#error("unsupported os")
+#endif
+
+fileprivate extension ByteBuffer {
+    mutating func readAndCompareString<Length: FixedWidthInteger>(to string: inout Substring, length: Length) -> Bool {
```
Comment (adam-fowler) [inline Sources/Hummingbird/Router/BinaryTrie/BinaryTrie+resolve.swift:223]: This shouldn't require an `inout Substring`. Use `withContiguousStorageIfAvailable` instead and if it fails (which is rare) copy string and call `makeContiguousUTF8` on copy and then call function with the string copy

**Review by adam-fowler — COMMENTED** [review]
Here are a series of initial comments. These should make the code more readable and more performant

File: Sources/Hummingbird/Router/BinaryTrie/BinaryTrie+resolve.swift [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+            let _token: BinaryTrieTokenKind.RawValue = readInteger(),
+            let token = BinaryTrieTokenKind(rawValue: _token)
+        else {
+            return nil
+        }
+
+        return token
+    }
+
+    mutating func readBinaryTrieNode() -> BinaryTrieNode? {
```
Comment (adam-fowler) [inline Sources/Hummingbird/Router/BinaryTrie/BinaryTrie+resolve.swift:283]: This function doesn't compile and it doesn't seem to be used

File: Sources/Hummingbird/Router/BinaryTrie/BinaryTrie+resolve.swift [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+        components: inout [Substring],
+        isInRecursiveWildcard: Bool
+    ) -> (value: Value, parameters: Parameters)? {
+        // If there are no more components in the path, return the value found
+        if components.isEmpty {
+            return value(for: index, parameters: parameters)
+        }
+
+        // Take the next component from the path
+        var component = components.removeFirst()
```
[the remark this answers is not in the document]
Comment (adam-fowler) [reply Sources/Hummingbird/Router/BinaryTrie/BinaryTrie+resolve.swift:146]: This is still using `removeFirst` is this intentional?

File: Sources/Hummingbird/Router/BinaryTrie/BinaryTrie+resolve.swift [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+                        parameters: &parameters,
+                        components: &components,
+                        isInRecursiveWildcard: false
+                    )
+                case .mismatch where isInRecursiveWildcard:
+                    if components.isEmpty {
+                        return nil
+                    }
+
+                    component = components[components.startIndex]
```
Comment (adam-fowler) [inline Sources/Hummingbird/Router/BinaryTrie/BinaryTrie+resolve.swift:176]: Can we guarantee at this point that the array is non-empty?

Comment (adam-fowler) [reply Sources/Hummingbird/Router/BinaryTrie/BinaryTrie+resolve.swift:176]: haha. just saw the check above, perhaps you could change this to
```swift
guard let component = components.first else { return nil }
```

File: Sources/Hummingbird/Router/TrieRouter.swift [hunk 6]
```diff
@@ -93,20 +93,62 @@ struct RouterPathTrieBuilder<Value: Sendable> {
     }
 }
 
-/// Trie used by Router responder
-struct RouterPathTrie<Value: Sendable>: Sendable {
-    let root: Node
+import NIOCore
+
+public struct BinaryRouterResponder<Context: BaseRequestContext>: HTTPResponder {
+    let trie: BinaryTrie<EndpointResponders<Context>>
```
Comment (adam-fowler) [inline Sources/Hummingbird/Router/TrieRouter.swift:99]: This seems like an odd place to add this type. Might be worthwhile adding its own file. The original `RouterResponder` has its own file.

File: Sources/Hummingbird/Router/Router.swift [hunk 7]
```diff
@@ -68,13 +68,13 @@ public final class Router<Context: BaseRequestContext>: RouterMethods, HTTPRespo
...
     /// build responder from router
-    public func buildResponder() -> RouterResponder<Context> {
+    public func buildResponder() -> BinaryRouterResponder<Context> {
         if self.options.contains(.autoGenerateHeadEndpoints) {
             self.trie.forEach { node in
                 node.value?.autoGenerateHeadEndpoint()
             }
         }
-        return RouterResponder(
+        return try! .init(
```
Comment (adam-fowler) [inline Sources/Hummingbird/Router/Router.swift:77]: `!`? We should remove this.

File: Sources/Hummingbird/Router/BinaryTrie/BinaryTrie+serialize.swift [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+
+        // Serialize the node's component
+        switch node.key {
+        case .path(let path):
+            trie.writeInteger(TokenKind.path.rawValue)
+            nextNodeOffsetIndex = reserveUInt32()
+
+            // Serialize the path constant
+            try trie.writeLengthPrefixed(as: Integer.self) { buffer in
+                buffer.writeSubstring(path)
```
[the remark this answers is not in the document]
Comment (adam-fowler) [reply Sources/Hummingbird/Router/BinaryTrie/BinaryTrie+serialize.swift:44]: In actual fact you could fix the issue where you are throwing an error from `init` and provide more detailing error information. 
```swift
func writeLengthPrefixedString(_ string: SubString) {
    do {
        try trie.writeLengthPrefixed(as: Integer.self) { buffer in
            buffer.writeSubstring(path)
        }
    } catch {
        preconditionFailure("Failed to write \"\(string)\" into BinaryTrie")
    }
}

[the remark this answers is not in the document]
Comment (adam-fowler) [reply Sources/Hummingbird/Router/BinaryTrie/BinaryTrie+serialize.swift:44]: Also have you checked performance if you use a `UInt16`. 64 bit integer is quite a bit of overkill here

**Review by adam-fowler — APPROVED** [review]
With the changes from #436 this is ready

**Pull request merged**